openscad.scm

```diff
@@ -0,0 +1,397 @@
+  "]" @prepend_indent_end @prepend_empty_softline @prepend_antispace
+)
+; to avoid having a list that directly follows a let_expression look visually unindented,
+; add another level of indentation:
+; let (
+;   u = true,
+; )[
+;  [1, 0, 0, 0],
+;  [0, 1, 0, 0],
+;  [0, 0, 1, 0],
+;  [0, 0, 0, 1],
+; ];
+(let_expression
+  (list
+    "[" @append_indent_start
+    "]" @prepend_indent_end
+  )
+)
+
+(range
+  "[" @append_antispace
+  "]" @prepend_antispace
+)
+(assignments "," @append_spaced_softline . [(block_comment) (line_comment)]* @do_nothing)
+(parameters "," @append_spaced_softline . [(block_comment) (line_comment)]* @do_nothing)
+
+; Never put a space before a comma semicolon or period/full stop
+[
+  ","
+  ";"
+  "."
+] @prepend_antispace
+
+; Don't insert spaces between the operator and their expression operand
+; '-x' v.s. '- x'
+(unary_expression _ @append_antispace . (expression))
+
+; ================================================================================
+; functions & modules
+; ================================================================================
+(function_item "function" @append_space)
+
+; indent the body of a function
+(function_item
+  (parameters)
+  .
+  "=" @append_spaced_softline @append_indent_start
+  (expression)
+  ";" @prepend_indent_end
+)
+
+; ===
+; function literals
+; ===
+(function_lit (parameters) @append_spaced_softline)
+(function_call (parenthesized_expression) @append_antispace (arguments))
+
+; ===
+; module_item, module_call, and transform_chain
+; ===
+
+; everything except `union_block` after a for/if/else statement should be a spaced_softline
+(module_item
+  body: [
+    (for_block)
+    (intersection_for_block)
+    (if_block)
+    (let_block)
+    (assign_block)
+    (transform_chain)
+    (include_statement)
+    (assert_statement)
+  ] @prepend_indent_start @append_indent_end
+)
+
+; module calls in a transformation chain will follow each other
+; sometimes staying on the same line and sometimes having a linebreak,
+; each linebreak typically also starts an indent scope
+(transform_chain) @prepend_input_softline
+(transform_chain
+  (modifier)*
+  (module_call) @append_indent_start
+  (transform_chain) @append_indent_end
+)
+
+; ================================================================================
+; blocks/expressions/statements
+; ================================================================================
+
+; Let child nodes handle indentation
+(var_declaration . (assignment . (identifier) . "=" @append_input_softline))
+
+(assignments) @append_space
+
+; NOTE uncomment for OpenSCAD 2022 and later
+; OS2021 does not support trailing commas in assignments
+; https://github.com/Leathong/openscad-LSP/issues/51#issuecomment-2891821939
+; (assignments
+;   (#delimiter! ",")
+;   (assignment) @append_delimiter
+;   .
+;   ","? @do_nothing
+;   .
+;   [(block_comment) (line_comment)]*
+;   .
+;   ")"
+;   .
+;   (#multi_line_only!)
+; )
```

Comment:
OpenSCAD 2022 can uncomment these lines. We can eventually split these up into a separate file and concatenate with something akin to `--latest-grammar`